plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -234,6 +262,130 @@ public void onFailure(Exception e) {
+                                + Arrays.toString(searchPipelineIds.toArray(new String[0])),
+                            RestStatus.CONFLICT
+                        )
+                    );
+
+            }
+
+        }, e -> {
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+
+    }
+
+    private void checkSearchPipelineBeforeDeleteModel(String modelId, ActionListener<Boolean> actionListener) {
+        GetSearchPipelineRequest getSearchPipelineRequest = new GetSearchPipelineRequest();
+        client.execute(GetSearchPipelineAction.INSTANCE, getSearchPipelineRequest, ActionListener.wrap(searchPipelineResponse -> {
+            if (!isPipelineContainsModel(
+                searchPipelineResponse.pipelines(),
+                modelId,
+                org.opensearch.search.pipeline.PipelineConfiguration::getConfigAsMap
+            )) {
+                actionListener.onResponse(true);
+            } else {
+                List<String> searchPipelineIds = getAllPipelineIds(
+                    searchPipelineResponse.pipelines(),
+                    org.opensearch.search.pipeline.PipelineConfiguration::getId
+                );
+                actionListener
+                    .onFailure(
+                        new OpenSearchStatusException(
+                            searchPipelineIds.size()
+                                + " search pipelines are still using this model, please delete or update the pipelines first: "
+                                + Arrays.toString(searchPipelineIds.toArray(new String[0])),
+                            RestStatus.CONFLICT
+                        )
+                    );
+
+            }
+
+        }, e -> {
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+
+    }
+
+    private void checkDownstreamTaskBeforeDeleteModel(String modelId, ActionListener<DeleteResponse> actionListener) {
+        CountDownLatch countDownLatch = new CountDownLatch(3);
+        AtomicBoolean noneBlocked = new AtomicBoolean(true);
+        ActionListener<Boolean> countDownActionListener = ActionListener.wrap(b -> {
+            countDownLatch.countDown();
+            noneBlocked.compareAndSet(true, b);
+            if (countDownLatch.getCount() == 0) {
+                if (noneBlocked.get()) {
+                    doDeleteModel(modelId, actionListener);
+                }
+            }
+        }, e -> {
+            countDownLatch.countDown();
+            noneBlocked.compareAndSet(true, false);
+            actionListener.onFailure(e);
+
+        });
+        checkAgentBeforeDeleteModel(modelId, countDownActionListener);
+        checkIngestPipelineBeforeDeleteModel(modelId, countDownActionListener);
+        checkSearchPipelineBeforeDeleteModel(modelId, countDownActionListener);
```

Comment:
Two questions:
1. Can you do these 3 operations in parallel?
2. If there is both an agent and an ingest pipeline depending on this model, will you only return the agents? It would better if you can return all the agents, ingest pipeline and search pipelines, so that the user would only meet with this error message once.

Reply:
1. Great work!
2. I mean, you can format the error message from 3 strings into 1. It's just low code effort but benefits many customers.

Reply:
For 2, the listener will multiply error message and I'm not sure whether we could queue them.

Reply:
I see. You can use multiple listener. Once all of them has received a response or error, then concatenate the message.

Reply:
Done. Now concate the messages with comma, and it would be result like "**1 search pipelines are still using this model, please delete or update the pipelines first: [rag_pipeline],1 agents are still using this model, please delete or update the agents first: [EyoCQ5MBi0GstA7LbauT]"**

Reply:
Great! Please add 1 unit test.